consensus/ibft/fork/manager.go

```diff
@@ -0,0 +1,434 @@
+	}
+
+	return signer, nil
+}
+
+// GetValidatorStore returns a proper validator set at specified height
+func (m *forkManagerImpl) GetValidatorStore(height uint64) (store.ValidatorStore, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, ok := m.validatorSets[ibftTypesToSourceType[fork.Type]]
+	if !ok {
+		return nil, ErrValidatorStoreNotFound
+	}
+
+	return set, nil
+}
+
+// GetValidators returns validators at specified height
+func (m *forkManagerImpl) GetValidators(height uint64) (validators.Validators, error) {
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	set, err := m.GetValidatorStore(height)
+	if err != nil {
+		return nil, err
+	}
+
+	return set.GetValidators(
+		calculateFetchingValidatorsHeight(height, m.epochSize, fork),
+	)
+}
+
+// GetHooks returns a hooks at specified height
+func (m *forkManagerImpl) GetHooks(height uint64) (hook.Hooks, error) {
+	hooks := &hook.HookManager{}
+
+	fork := m.getFork(height)
+	if fork == nil {
+		return nil, ErrForkNotFound
+	}
+
+	var err error
+
+	switch fork.Type {
+	case PoA:
+		err = m.registerPoAHooks(hooks, height)
+	case PoS:
+		registerPoSHook(hooks, m.epochSize)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	if err := m.registerPoAPrepareHooks(hooks, height); err != nil {
+		return nil, err
+	}
+
+	m.registerPoSPrepareHooks(hooks, height)
+
+	return hooks, nil
+}
+
+// Close calls termination process of submodules
+func (m *forkManagerImpl) Close() error {
+	if err := m.closeSnapshotValidatorStore(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// initializeSigners initialize all signers based on Fork configuration
+func (m *forkManagerImpl) initializeSigners() error {
+	for _, fork := range m.forks {
+		valType := fork.ValidatorType
+
+		if err := m.initializeSigner(valType); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initializeValidatorStores initializes all validator sets based on Fork configuration
+func (m *forkManagerImpl) initializeValidatorStores() error {
+	for _, fork := range m.forks {
+		sourceType := ibftTypesToSourceType[fork.Type]
+		if err := m.initializeValidatorStore(sourceType); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// initializeSigner initializes the specified signer
+func (m *forkManagerImpl) initializeSigner(valType validators.ValidatorType) error {
+	if _, ok := m.signers[valType]; ok {
+		return nil
+	}
+
+	signer, err := signer.NewSignerFromType(m.secretsManager, valType)
+	if err != nil {
+		return err
+	}
+
+	m.signers[valType] = signer
+
+	return nil
+}
+
+// initializeValidatorStore initializes the specified validator set
+func (m *forkManagerImpl) initializeValidatorStore(setType store.SourceType) error {
+	if _, ok := m.validatorSets[setType]; ok {
+		return nil
+	}
+
+	var (
+		valSet store.ValidatorStore
+		err    error
+	)
+
+	switch setType {
+	case store.Snapshot:
+		valSet, err = m.initializeSnapshotValidatorStore()
```

Comment:
The logic in `initializeSnapshotValidatorStore` seems like it belongs to the `snapshot` package. Most of the initial logic is reading a file from disk. Consider changing the construction of a snapshot store to be consistent with contract store creation:

```
snapshot.NewStore(args...)
```

Reply:
To be honest, I don't want Snapshot store to have side effect logic (reading/writing data from/to file)

Reply:
@dbrajovic What's your suggestion?

Reply:
[Closed](https://github.com/0xPolygon/polygon-edge/pull/649#pullrequestreview-1084070770)